### DISCUSSION TITLE
Project Roles / User Groups

### DISCUSSION THREAD

**Discussion by codyebberson** [medplum/medplum#1392] [Ideas]
# Project Roles / User Groups

At present, the notion of "role" (as in "Role Based Access Controls" or "RBAC") is implicit using `AccessPolicy` and `UserConfiguration`.  The current data model provides the necessary building blocks to create complex role-based systems without a formal representation of "role".

The problem: It is unclear how to map `AccessPolicy` and `UserConfiguration` to the a traditional enterprise "role" or "group" as defined by SCIM, Okta, or other enterprise user management tools.

## Goals

* Identify a solution to allow user management via SCIM and Okta
* Project Admins should be able to add/remove users using the Okta admin tool via SCIM
* Project Admins should be able to add/remove users from groups using the Okta admin tool via SCIM

## Proposal

* Introduce a new `ProjectRole` resource type
* `ProjectRole` has the following properties
  * `name`
  * `accessPolicy`
  * `userConfiguration`
* Replace `ProjectMembership.accessPolicy` and `ProjectMembership.userConfiguration` with `ProjectMembership.role`

## Questions

* Name?
  * `ProjectRole`
  * `ProjectGroup`
  * `UserRole`
  * `UserGroup`
  * Cannot use `Group`, because that is already a FHIR resource name, and I would prefer not to introduce ambiguity there
* Should `members` be an array field on the `ProjectRole` type?
  * Pros: more aligned with SCIM object model
  * Cons: more cumbersome to work with, especially in large organizations (1000+ members)
* Can a user be a member of multiple groups at the same time?
  * SCIM allows users to be in multiple groups
  * What would that mean for `AccessPolicy`?
  * Would the resulting `AccessPolicy` be the union of all groups?
* How does this interact with SMART scopes?
* Should `UserConfiguration` be in the role?  Or is it best left on the project membership?
  * In the common case, it's faster and easier if everything is in one "package", so that a project admin can setup a user in one step
  * It makes per-user customization harder though
  * We could allow for "use the role UserConfiguration by default, unless a user specifies an override"

## References

SCIM Object Model
https://www.simplecloud.info/

SCIM Group Schema
https://www.simplecloud.info/specs/draft-scim-core-schema-01.html#group-resource

Okta User Management
https://help.okta.com/en-us/Content/Topics/users-groups-profiles/usgp-main.htm

Okta Groups
https://help.okta.com/en-us/Content/Topics/users-groups-profiles/usgp-about-groups.htm

Okta - Assign user to a group
https://help.okta.com/en-us/Content/Topics/users-groups-profiles/usgp-assign-group-people.htm

Microsoft 365 User Management
https://learn.microsoft.com/en-us/microsoft-365/enterprise/deploy-identity-solution-overview

Microsoft 365 Groups
https://support.microsoft.com/en-us/office/learn-about-microsoft-365-groups-b565caa1-5c40-40ef-9915-60fdb2d97fa2

**Comment by reshmakh** [comment]
In the spirit of standards compliance and trying to use commonly used language to describe concepts, my thoughts on the 4 proposals.

1. I think `ProjectGroup` or perhaps`UserGroup` is probably the most clear and in the spirit of what we are trying to convey.  It's also probable that we will end up with groups that are above the project over time, especially for self-hosted 
2. I do not think members as an array field is a good pattern, especially as programmatic access/manipulation will be a common case here.  
3. A concept of "Primary" for the group to determine the AccessPolicy seems like a useful concept.  We should also leave open the possibility of hierarchy for groups.
4. SMART Scopes filter seems like the compliant approach given the current use cases
5. UserConfiguration at the project level makes sense to me.  In practice, the person designing the UserConfiguration and the user management will often not be the same person.  A sane default is useful.
